File: app/src/main/java/ch/bailu/aat/preferences/SolidDirectoryQuery.java
package ch.bailu.aat.preferences;

import android.content.Context;

import ch.bailu.aat.R;
import ch.bailu.aat.services.directory.GpxDbConstants;

public class SolidDirectoryQuery extends SolidString {
    private static final String KEY_DIR_DIRECTORY="DIR_DIRECTORY";
    private static final String KEY_DIR_INDEX="DIR_INDEX_";

    private final static String KEY_DATE_START="DATE_START_";
    private final static String KEY_DATE_END="DATE_END_";
    private final static String KEY_USE_DATE_START="USE_DATE_START_";
    private final static String KEY_USE_DATE_END="USE_DATE_END_";

    private final static String KEY_USE_GEO="USE_GEO_";

    private final static String KEY_BOUNDING_BOX = "BOX_";
    

    public SolidDirectoryQuery(Context c) {
        super(Storage.global(c), KEY_DIR_DIRECTORY);

    }

    public boolean containsKey(String s) {
        return s.contains(getValueAsString());
    }
    
    public SolidInteger getPosition() {
        return new SolidInteger(getStorage(), KEY_DIR_INDEX+ getValueAsString());
    }


    public SolidBoolean getUseDateStart() {
        return new SolidBoolean(getStorage(), KEY_USE_DATE_START+ getValueAsString());
    }


    public SolidBoolean getUseDateEnd() {
        return new SolidBoolean(getStorage(), KEY_USE_DATE_END+ getValueAsString());
    }


    public SolidDate getDateStart() {
        return new SolidDate(
                getStorage(),
                KEY_DATE_START+ getValueAsString(),
                getContext().getString(R.string.filter_date_start));

    }


    public SolidDate getDateTo() {
        return new SolidDate(
                getStorage(),
                KEY_DATE_END+ getValueAsString(),
                getContext().getString(R.string.filter_date_to));
    }

    public SolidBoolean getUseGeo() {
        return new SolidBoolean(getStorage(), KEY_USE_GEO+ getValueAsString());

    }

    public SolidBoundingBox getBoundingBox() {
        return new SolidBoundingBox(
                getStorage(),
                KEY_BOUNDING_BOX+ getValueAsString(),
                getContext().getString(R.string.filter_geo));
    }

    
    
    public String createSelectionString() {
        final String d = createSelectionStringDate();
        final String b = createSelectionStringBounding(); 
        
        if (d.length()>0 && b.length()>0) {
            return "(" + d + ") AND (" + b + ")";
        } else {
            return d + b;
        }
    }
    
    
    public String createSelectionStringBounding() {
        if (getUseGeo().getValue()) {
            return getBoundingBox().createSelectionStringInside();
        }
        
        return "";
    }
    
    
    private final static long DAY = 1000*60*60*24; // ms*sec*min*h = d
    public String createSelectionStringDate() {
        String selection="";


        if (getUseDateStart().getValue() || getUseDateEnd().getValue()) {
            long start, end;

            if (getUseDateStart().getValue()) {
                start = getDateStart().getValue();
            } else {
                start = 0;
            }

            if (getUseDateEnd().getValue()) {
                end = getDateTo().getValue();
            } else {
                end = System.currentTimeMillis()/DAY;
                end = end + 5;
                end = end * DAY;
            }


            selection = GpxDbConstants.KEY_START_TIME 
                    + " BETWEEN " 
                    + String.valueOf(Math.min(start, end)) 
                    + " AND " 
                    + String.valueOf(Math.max(start, end)) ;
        }

        return selection;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/file_list/SolidDirectoryQuery.kt
package ch.bailu.aat_lib.preferences.file_list

import ch.bailu.aat_lib.preferences.SolidBoolean
import ch.bailu.aat_lib.preferences.SolidBoundingBox
import ch.bailu.aat_lib.preferences.SolidDate
import ch.bailu.aat_lib.preferences.SolidFile
import ch.bailu.aat_lib.preferences.SolidInteger
import ch.bailu.aat_lib.preferences.SolidString
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.resources.Res
import ch.bailu.aat_lib.service.directory.database.GpxDbConfiguration
import ch.bailu.foc.FocFactory

class SolidDirectoryQuery(storage: StorageInterface, focFactory: FocFactory) : SolidFile(storage, KEY_DIR_DIRECTORY, focFactory) {

    fun containsKey(key: String): Boolean {
        return key.contains(getValueAsString())
    }

    val position: SolidInteger
        get() = SolidInteger(getStorage(), KEY_DIR_INDEX + getValueAsString())

    val useDateStart: SolidBoolean
        get() = SolidBoolean(getStorage(), KEY_USE_DATE_START + getValueAsString())

    val useDateEnd: SolidBoolean
        get() = SolidBoolean(getStorage(), KEY_USE_DATE_END + getValueAsString())

    val dateStart: SolidDate
        get() = SolidDate(
            getStorage(),
            KEY_DATE_START + getValueAsString(),
            Res.str().filter_date_start()
        )

    val dateEnd: SolidDate
        get() = SolidDate(
            getStorage(),
            KEY_DATE_END + getValueAsString(),
            Res.str().filter_date_to()
        )

    val useGeo: SolidBoolean
        get() = SolidBoolean(getStorage(), KEY_USE_GEO + getValueAsString())

    val boundingBox: SolidBoundingBox
        get() = SolidBoundingBox(
            getStorage(),
            KEY_BOUNDING_BOX + getValueAsString(),
            Res.str().filter_geo()
        )

    val solidNameFilter: SolidString
        get() = SolidString(getStorage(), KEY_NAME_FILTER + getValueAsString())

    val solidSortAttribute: SolidSortAttribute
        get() = SolidSortAttribute(getStorage(), KEY_SORT_ATTRIBUTE + getValueAsString())

    val solidSortOrderAscend: SolidBoolean
        get() = SolidSortOrderAscend(getStorage(), KEY_SORT_ASCEND + getValueAsString())

    fun isFilterEnabled(): Boolean {
        val filterName = solidNameFilter.getValueAsStringNonDef()
        return useGeo.isEnabled || useDateEnd.isEnabled || useDateStart.isEnabled || filterName.trim().isNotEmpty()
    }

    fun createExtraStatement(): String {
        var delimiter = " WHERE "
        val builder = StringBuilder()
        arrayOf(createNameFilterStatement(), createDateFilterStatement(), createBoundingFilterStatement()).forEach {
            if (it.isNotEmpty()) {
                builder.append(delimiter)
                builder.append(it)
                delimiter = " AND "
            }
        }
        builder.append(createSortStatement())
        return builder.toString()
    }

    fun createExtraStatementParam(): String {
        val filterName = solidNameFilter.getValueAsStringNonDef().trim()
        if (filterName.isNotEmpty()) {
            return "%$filterName%"
        }
        return ""
    }

    private fun createSortStatement(): String {
        val direction = if (solidSortOrderAscend.value) { "ASC" } else { "DESC" }
        return " ORDER BY ${solidSortAttribute.getAttributeName()} $direction"
    }
    private fun createNameFilterStatement(): String {
        val nameFilter = solidNameFilter.getValueAsStringNonDef().trim()
        return if (nameFilter.isNotEmpty()) {
            "${GpxDbConfiguration.ATTR_FILENAME} like ?"
        } else {
            ""
        }
    }

    private fun createBoundingFilterStatement(): String {
        return if (useGeo.value) {
            boundingBox.createSelectionStringInside()
        } else ""
    }

    private fun createDateFilterStatement(): String {
        var selection = ""
        if (useDateStart.value || useDateEnd.value) {
            var end: Long
            val start: Long = if (useDateStart.value) {
                dateStart.getValue()
            } else {
                0
            }
            if (useDateEnd.value) {
                end = dateEnd.getValue()
            } else {
                end = System.currentTimeMillis() / DAY
                end += 5
                end *= DAY
            }
            selection = (GpxDbConfiguration.ATTR_START_TIME
                    + " BETWEEN "
                    + Math.min(start, end)
                    + " AND "
                    + Math.max(start, end))
        }
        return selection
    }

    companion object {
        private const val KEY_DIR_DIRECTORY = "DIR_DIRECTORY"
        private const val KEY_DIR_INDEX = "DIR_INDEX_"
        private const val KEY_DATE_START = "DATE_START_"
        private const val KEY_DATE_END = "DATE_END_"
        private const val KEY_USE_DATE_START = "USE_DATE_START_"
        private const val KEY_USE_DATE_END = "USE_DATE_END_"
        private const val KEY_USE_GEO = "USE_GEO_"
        private const val KEY_BOUNDING_BOX = "BOX_"
        private const val KEY_NAME_FILTER = "NAME_FILTER_"
        private const val KEY_SORT_ATTRIBUTE = "SORT_ATTRIBUTE_"
        private const val KEY_SORT_ASCEND = "SORT_ASCEND_"
        private const val DAY = (1000 * 60 * 60 * 24).toLong() // /* ms*sec*min*h = d
    }
}
